domain/blockdag/block_utxo.go

```diff
@@ -146,27 +140,69 @@ func (node *blockNode) applyBlueBlocks(selectedParentPastUTXO UTXOSet, blueBlock
 		}
 		isSelectedParent := i == 0
 
-		for j, tx := range blueBlock.Transactions() {
+		for j, tx := range transactions {
 			var isAccepted bool
+			var txFee uint64
 
-			// Coinbase transaction outputs are added to the UTXO
-			// only if they are in the selected parent chain.
-			if !isSelectedParent && tx.IsCoinBase() {
-				isAccepted = false
-			} else {
-				isAccepted, err = pastUTXO.AddTx(tx.MsgTx(), node.blueScore)
-				if err != nil {
-					return nil, nil, err
-				}
+			isAccepted, txFee, accumulatedMass, err =
+				node.maybeAcceptTx(tx, isSelectedParent, pastUTXO, accumulatedMass, selectedParentMedianTime)
+			if err != nil {
+				return nil, nil, err
 			}
-			blockTxsAcceptanceData.TxAcceptanceData[j] = TxAcceptanceData{Tx: tx, IsAccepted: isAccepted}
+
+			blockTxsAcceptanceData.TxAcceptanceData[j] = TxAcceptanceData{
+				Tx:         tx,
+				Fee:        txFee,
+				IsAccepted: isAccepted}
 		}
 		multiBlockTxsAcceptanceData[i] = blockTxsAcceptanceData
 	}
 
 	return pastUTXO, multiBlockTxsAcceptanceData, nil
 }
 
+func (node *blockNode) maybeAcceptTx(tx *util.Tx, isSelectedParent bool, pastUTXO UTXOSet,
+	accumulatedMassBefore uint64, selectedParentMedianTime mstime.Time) (
+	isAccepted bool, txFee uint64, accumulatedMassAfter uint64, err error) {
+
+	accumulatedMass := accumulatedMassBefore
+
+	// Coinbase transaction outputs are added to the UTXO-set only if they are in the selected parent chain.
+	if tx.IsCoinBase() {
+		if !isSelectedParent {
+			return false, 0, 0, nil
+		}
+		txMass := CalcTxMass(tx, nil)
+		accumulatedMass += txMass
+
+		_, err = pastUTXO.AddTx(tx.MsgTx(), node.blueScore)
+		if err != nil {
+			return false, 0, 0, err
+		}
+
+		return true, 0, accumulatedMass, nil
+	}
+
+	txFee, accumulatedMassAfter, err = node.dag.checkConnectTransactionToPastUTXO(
+		node, tx, pastUTXO, accumulatedMassBefore, selectedParentMedianTime)
+	if err != nil {
+		if !errors.As(err, &(RuleError{})) {
```

Comment:
This will have to modify quite a few unrelated places, and I don't want to enhance this monster of PR any further, so created ticket:
https://daglabs.atlassian.net/browse/NOD-1366